Tests/SwiftParserTest/ExpressionTests.swift

```diff
@@ -2742,4 +2742,88 @@ final class StatementExpressionTests: ParserTestCase {
       """
     )
   }
+
+  func testTypedThrowsDisambiguation() {
+    assertParse(
+      "[() throws(MyError) 1️⃣async -> Void]()",
+      diagnostics: [
+        DiagnosticSpec(message: "'async' must precede 'throws'", fixIts: ["move 'async' in front of 'throws'"])
+      ],
+      fixedSource: "[() async throws(MyError) -> Void]()",
+      experimentalFeatures: .typedThrows
+    )
+    assertParse(
+      "[() throws 1️⃣async2️⃣(MyError) -> Void]()",
+      diagnostics: [
+        DiagnosticSpec(locationMarker: "1️⃣", message: "'async' must precede 'throws'", fixIts: ["move 'async' in front of 'throws'"]),
+        DiagnosticSpec(locationMarker: "2️⃣", message: "unexpected code '(MyError)' in array element"),
+      ],
+      fixedSource: "[() async throws (MyError) -> Void]()",
+      experimentalFeatures: .typedThrows
+    )
+    assertParse(
+      "[() try(MyError) async -> Void]()",
```

Comment:
This should not parse. This seems to be an existing issue in the new parser. I filed https://github.com/apple/swift-syntax/issues/2255